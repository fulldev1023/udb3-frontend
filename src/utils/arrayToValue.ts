const arrayToValue = <T>(value: T[] | T): T =>
  Array.isArray(value) ? value.at(0) : value;

export { arrayToValue };
