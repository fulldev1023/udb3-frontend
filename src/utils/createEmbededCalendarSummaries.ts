type TextFormat = 'xs-text' | 'sm-text' | 'md-text' | 'lg-text';
type HtmlFormat = 'xs-hmtl' | 'sm-html' | 'md-html' | 'lg-html';
type Format = TextFormat | HtmlFormat;

type CalendarSummaries = {
  [key: string]: Format;
};

const createEmbededCalendarSummaries = (formats: Format[]): CalendarSummaries =>
  formats.reduce(
    (accumulator, currentValue, currentIndex) => ({
      ...accumulator,
      [`embedCalendarSummaries[${currentIndex}]`]: currentValue,
    }),
    {},
  );

export { createEmbededCalendarSummaries };
export type { Format as CalendarSummaryFormat };
