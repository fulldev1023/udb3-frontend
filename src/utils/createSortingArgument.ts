const createSortingArgument = (sortOptions: {
  field: string;
  order: string;
}) => {
  return { [`sort[${sortOptions.field}]`]: `${sortOptions.order}` };
};

export { createSortingArgument };
