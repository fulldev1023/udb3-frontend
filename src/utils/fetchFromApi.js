import getConfig from 'next/config';

const Errors = {
  401: 'unauthorised',
  403: 'forbidden',
  204: 'no_content',
  404: 'bad_request',
};

const fetchFromApi = async ({ path, searchParams = {}, options = {} }) => {
  const { publicRuntimeConfig } = getConfig();
  const url = new URL(`${publicRuntimeConfig.apiUrl}${path}`);
  url.search = new URLSearchParams(searchParams);
  const response = await fetch(url, options);

  if (Object.keys(Errors).includes(response.status)) {
    throw new Error(Errors[response.status]);
  }

  return response;
};

export { fetchFromApi, Errors };
