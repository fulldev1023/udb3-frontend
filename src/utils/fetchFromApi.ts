import omit from 'lodash/omit';
import getConfig from 'next/config';

class FetchError extends Error {
  status: number;

  constructor(status: number, message: string) {
    super(message);
    this.status = status;
  }
}

type ErrorObject = {
  type: 'ERROR';
  status?: number;
  message: string;
};

const isErrorObject = (value: any): value is ErrorObject => {
  return (
    value.type === 'ERROR' &&
    typeof value.message === 'string' &&
    (!value.status || typeof value.status === 'number')
  );
};

const fetchFromApi = async ({
  path,
  searchParams = {},
  options = {},
  silentError = false,
}): Promise<ErrorObject | Response> => {
  const { publicRuntimeConfig } = getConfig();

  // eslint-disable-next-line no-console
  console.log({ publicRuntimeConfig });

  let response: Response;
  let url: URL;

  try {
    url = new URL(`${publicRuntimeConfig.apiUrl}${path}`);
    url.search = new URLSearchParams(omit(searchParams, 'queryKey')).toString();
  } catch (e) {
    if (!silentError) {
      throw new FetchError(400, e?.message ?? 'Unknown error');
    }
    return {
      type: 'ERROR',
      message: e.message ?? 'Unknown error',
    };
  }

  try {
    response = await fetch(url.toString(), options);
  } catch (e) {
    if (!silentError) {
      throw new FetchError(response?.status, e?.message ?? 'Unknown error');
    }

    return {
      type: 'ERROR',
      message: e.message ?? 'Unknown error',
    };
  }

  if (!response.ok) {
    const result = await response.json();

    if (!silentError) {
      throw new FetchError(response?.status, result?.title ?? 'Unknown error');
    }

    return {
      type: 'ERROR',
      status: response?.status,
      message: result?.title ?? 'Unknown error',
    };
  }

  return response;
};

export { FetchError, fetchFromApi, isErrorObject };
export type { ErrorObject };
