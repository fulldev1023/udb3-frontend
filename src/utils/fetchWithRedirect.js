const Errors = {
  UNAUTHORIZED: 'unauthorized',
};

const fetchWithRedirect = async (...args) => {
  const response = await fetch(...args);
  if (response.status === 401 || response.status === 403) {
    throw Error(Errors.UNAUTHORIZED);
  }
  return response;
};

export { fetchWithRedirect, Errors };
