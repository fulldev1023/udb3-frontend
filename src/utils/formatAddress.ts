import type { AddressInternal } from '@/types/Address';

const formatAddressInternal = (addressInternal: AddressInternal) => {
  return `${addressInternal.streetAddress} ${addressInternal.postalCode} ${addressInternal.addressLocality}`;
};

export { formatAddressInternal };
