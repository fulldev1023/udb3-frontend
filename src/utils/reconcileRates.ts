import {
  PriceCategories,
  Rate,
} from '@/pages/steps/AdditionalInformationStep/PriceInformation';
import { Offer } from '@/types/Offer';
import { isUitpasOrganizer } from '@/pages/steps/AdditionalInformationStep/OrganizerPicker';

const parseNumber = (number: string | number): number => {
  return Number(String(number).replace(',', '.'));
};

export function reconcileRates(
  currentRates: Rate[],
  newRates: Rate[],
  offer?: Offer,
): Rate[] {
  const mainLanguage = offer?.mainLanguage;
  const formatter = Intl.NumberFormat('nl-NL', {
    style: 'decimal',
    minimumFractionDigits: 2,
    useGrouping: false,
  });

  const alreadyEmbeddedUitpasPrices = currentRates.some(
    (rate) => rate.category === PriceCategories.UITPAS,
  );

  const isDefaultValue =
    currentRates.length === 1 && currentRates[0].price === '';

  if (isDefaultValue && newRates.length) {
    currentRates = newRates;
  }

  const hasUitpasOrganizer =
    offer?.organizer && isUitpasOrganizer(offer?.organizer);

  if (alreadyEmbeddedUitpasPrices && (newRates.length || hasUitpasOrganizer)) {
    currentRates = currentRates.filter(
      (rate) => rate.category !== PriceCategories.UITPAS,
    );
  }

  if (hasUitpasOrganizer) {
    newRates.forEach((rate) => {
      if (rate.category === PriceCategories.UITPAS) {
        currentRates.push(rate);
      }
    });
  }

  currentRates = currentRates.map((rate) => {
    if (mainLanguage) {
      rate.name = {
        [mainLanguage]: Object.values(rate.name)[0],
        ...rate.name,
      };
    }

    return {
      ...rate,
      price: formatter.format(parseNumber(rate.price)),
    };
  });

  return currentRates;
}
