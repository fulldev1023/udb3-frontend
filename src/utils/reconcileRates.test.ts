import {
  PriceCategories,
  Rate,
} from '@/pages/steps/AdditionalInformationStep/PriceInformation';
import { Offer } from '@/types/Offer';
import { reconcileRates } from '@/utils/reconcileRates';

describe('reconcileRates', () => {
  it('can overwrite default values', () => {
    const currentRates: Rate[] = [
      {
        name: { nl: 'Base Tarif' },
        category: PriceCategories.BASE,
        price: '',
        priceCurrency: 'EUR',
      },
    ];

    const newRates = [
      {
        name: { nl: 'foobar' },
        category: PriceCategories.BASE,
        price: '15',
        priceCurrency: 'EUR',
      },
    ];

    expect(reconcileRates(currentRates, newRates)).toMatchSnapshot();
    expect(reconcileRates(currentRates, [])).toEqual([
      {
        name: { nl: 'Base Tarif' },
        category: PriceCategories.BASE,
        price: '0,00',
        priceCurrency: 'EUR',
      },
    ]);
  });

  it('can format prices', () => {
    const currentRates: Rate[] = [
      {
        name: { nl: 'foo' },
        category: PriceCategories.BASE,
        price: '5',
        priceCurrency: 'EUR',
      },
      {
        name: { nl: 'bar' },
        category: PriceCategories.BASE,
        price: 25,
        priceCurrency: 'EUR',
      },
    ];

    expect(reconcileRates(currentRates, [])).toEqual([
      {
        name: { nl: 'foo' },
        category: PriceCategories.BASE,
        price: '5,00',
        priceCurrency: 'EUR',
      },
      {
        name: { nl: 'bar' },
        category: PriceCategories.BASE,
        price: '25,00',
        priceCurrency: 'EUR',
      },
    ]);
  });

  it('can append embedded UiTPAS prices', () => {
    const currentRates: Rate[] = [
      {
        name: { nl: 'foobar' },
        category: PriceCategories.BASE,
        price: '20',
        priceCurrency: 'EUR',
      },
    ];

    const newRates = [
      {
        name: { nl: 'foobar' },
        category: PriceCategories.BASE,
        price: '15',
        priceCurrency: 'EUR',
      },
      {
        name: { nl: 'foobar' },
        category: PriceCategories.UITPAS,
        price: '25',
        priceCurrency: 'EUR',
      },
    ];

    expect(reconcileRates(currentRates, newRates)).toMatchSnapshot();
  });

  it('can updated embedded UiTPAS prices', () => {
    const currentRates: Rate[] = [
      {
        name: { nl: 'foo' },
        category: PriceCategories.BASE,
        price: '15',
        priceCurrency: 'EUR',
      },
      {
        name: { nl: 'bar' },
        category: PriceCategories.UITPAS,
        price: '25',
        priceCurrency: 'EUR',
      },
    ];

    const newRates = [
      {
        name: { nl: 'foo' },
        category: PriceCategories.BASE,
        price: '15',
        priceCurrency: 'EUR',
      },
      {
        name: { nl: 'bar' },
        category: PriceCategories.UITPAS,
        price: '20',
        priceCurrency: 'EUR',
      },
    ];

    expect(reconcileRates(currentRates, newRates)).toMatchSnapshot();
  });

  it('does not add back deleted prices', () => {
    const currentRates: Rate[] = [
      {
        name: { nl: 'foo' },
        category: PriceCategories.BASE,
        price: '20',
        priceCurrency: 'EUR',
      },
      {
        name: { nl: 'uitpas' },
        category: PriceCategories.UITPAS,
        price: '25',
        priceCurrency: 'EUR',
      },
    ];

    const newRates = [
      {
        name: { nl: 'foo' },
        category: PriceCategories.BASE,
        price: '20',
        priceCurrency: 'EUR',
      },
      {
        name: { nl: 'bar' },
        category: PriceCategories.BASE,
        price: '20',
        priceCurrency: 'EUR',
      },
      {
        name: { nl: 'uitpas' },
        category: PriceCategories.UITPAS,
        price: '20',
        priceCurrency: 'EUR',
      },
    ];

    expect(reconcileRates(currentRates, newRates)).toMatchSnapshot();
  });

  it('can backfill price languages', () => {
    const currentRates: Rate[] = [
      {
        name: { de: 'foo' },
        category: PriceCategories.BASE,
        price: '20',
        priceCurrency: 'EUR',
      },
      {
        name: { de: 'foo', fr: 'Le Foo' },
        category: PriceCategories.BASE,
        price: '20',
        priceCurrency: 'EUR',
      },
    ];

    const newRates = [
      {
        name: { nl: 'uitpas' },
        category: PriceCategories.UITPAS,
        price: '20',
        priceCurrency: 'EUR',
      },
    ];

    expect(
      reconcileRates(currentRates, newRates, { mainLanguage: 'fr' } as Offer),
    ).toEqual([
      {
        name: { fr: 'foo', de: 'foo' },
        category: PriceCategories.BASE,
        price: '20,00',
        priceCurrency: 'EUR',
      },
      {
        name: { fr: 'Le Foo', de: 'foo' },
        category: PriceCategories.BASE,
        price: '20,00',
        priceCurrency: 'EUR',
      },
      {
        name: { fr: 'uitpas', nl: 'uitpas' },
        category: PriceCategories.UITPAS,
        price: '20,00',
        priceCurrency: 'EUR',
      },
    ]);
  });
});
