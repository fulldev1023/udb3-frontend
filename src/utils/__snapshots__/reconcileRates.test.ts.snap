// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reconcileRates can append embedded UiTPAS prices 1`] = `
Array [
  Object {
    "category": "base",
    "name": Object {
      "nl": "foobar",
    },
    "price": "20,00",
    "priceCurrency": "EUR",
  },
  Object {
    "category": "uitpas",
    "name": Object {
      "nl": "foobar",
    },
    "price": "25,00",
    "priceCurrency": "EUR",
  },
]
`;

exports[`reconcileRates can overwrite default values 1`] = `
Array [
  Object {
    "category": "base",
    "name": Object {
      "nl": "foobar",
    },
    "price": "15,00",
    "priceCurrency": "EUR",
  },
]
`;

exports[`reconcileRates does not add back deleted prices 1`] = `
Array [
  Object {
    "category": "base",
    "name": Object {
      "nl": "foo",
    },
    "price": "20,00",
    "priceCurrency": "EUR",
  },
  Object {
    "category": "uitpas",
    "name": Object {
      "nl": "uitpas",
    },
    "price": "20,00",
    "priceCurrency": "EUR",
  },
]
`;

exports[`reconcileRates can updated embedded UiTPAS prices 1`] = `
Array [
  Object {
    "category": "base",
    "name": Object {
      "nl": "foo",
    },
    "price": "15,00",
    "priceCurrency": "EUR",
  },
  Object {
    "category": "uitpas",
    "name": Object {
      "nl": "bar",
    },
    "price": "20,00",
    "priceCurrency": "EUR",
  },
]
`;
